src/Stratis.Bitcoin/P2P/Peer/NetworkPeerClient.cs

```diff
@@ -0,0 +1,172 @@
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public NetworkPeerClient(int Id, TcpClient tcpClient, ILoggerFactory loggerFactory)
+        {
+            this.tcpClient = tcpClient;
+
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, string.Format("[{0}{1}] ", this.Id, this.RemoteEndPoint != null ? "-" + this.RemoteEndPoint.ToString() : ""));
+
+            this.Stream = this.tcpClient.Connected ? this.tcpClient.GetStream() : null;
+            this.ProcessingCompletion = new TaskCompletionSource<bool>();
+
+            this.writeLock = new AsyncLock();
+        }
+
+        /// <summary>
+        /// Connects the network client to the target server.
+        /// </summary>
+        /// <param name="endPoint">IP address and port to connect to.</param>
+        /// <param name="cancellation">Cancellation token that allows aborting the operation.</param>
+        /// <exception cref="OperationCanceledException">Thrown when the connection attempt was aborted.</exception>
+        public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation)
+        {
+            this.logger = this.loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.Id}-{endPoint}] ");
+            this.logger.LogTrace("({0}:'{1}')", nameof(endPoint), endPoint);
+
+            try
+            {
+                await Task.Run(async () => await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port), cancellation);
+                this.Stream = this.tcpClient.GetStream();
+            }
+            catch (Exception e)
+            {
+                if (e is OperationCanceledException)
+                {
+                    this.logger.LogTrace("Connecting to '{0}' cancelled.", endPoint);
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                }
+                else
+                {
+                    this.logger.LogDebug("Error connecting to '{0}', exception: {1}", endPoint, e.ToString());
+                    this.logger.LogTrace("(-)[UNHANDLED_EXCEPTION]");
+                }
+                throw;
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Sends data over the established connection.
+        /// </summary>
+        /// <param name="data">Data to send.</param>
+        /// <param name="cancellation">Cancellation token that allows aborting the operation.</param>
+        /// <exception cref="OperationCanceledException">Thrown when the connection was terminated or the cancellation token was cancelled.</exception>
+        public async Task SendAsync(byte[] data, CancellationToken cancellation)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(data), nameof(data.Length), data.Length);
+
+            using (await this.writeLock.LockAsync(cancellation))
```

Comment:
Come to think about it I am not sure a lock is required here, we will only push to this client messages sequentially from a blocking queue (unless you  have plan to change that)

Reply:
i changed that, yes, i completely removed the queue, not sure in this PR though :-/

Reply:
Awesome yes we let the scheduler be the queue.